src/goto-analyzer/unreachable_instructions.cpp

```diff
@@ -243,22 +354,89 @@ static void list_functions(
         json_result);
   }
 
-  if(json && !json_result.array.empty())
+  if(options.get_bool_option("json") && !json_result.array.empty())
     os << json_result << '\n';
+  else if(options.get_bool_option("xml"))
+    os << xml_result << '\n';
 }
 
 void unreachable_functions(
   const goto_modelt &goto_model,
   const bool json,
   std::ostream &os)
 {
-  list_functions(goto_model, json, os, true);
+  optionst options;
+  if (json)
+    options.set_option("json", true);
+  else
+    options.set_option("text", true);
+
+  std::set<irep_idt> called = compute_called_functions(goto_model);
+
+  list_functions(goto_model, called, options, os, true);
 }
 
 void reachable_functions(
   const goto_modelt &goto_model,
   const bool json,
   std::ostream &os)
 {
-  list_functions(goto_model, json, os, false);
+  optionst options;
+  if (json)
+    options.set_option("json", true);
+  else
+    options.set_option("text", true);
+
+  std::set<irep_idt> called = compute_called_functions(goto_model);
+
+  list_functions(goto_model, called, options, os, false);
+}
+
+
+std::set<irep_idt> compute_called_functions_from_ai(
+  const goto_modelt &goto_model,
+  const ai_baset &ai)
+{
+  std::set<irep_idt> called;
+
+  forall_goto_functions(f_it, goto_model.goto_functions)
+  {
+    if(!f_it->second.body_available())
+      continue;
+
+    const goto_programt &p = f_it->second.body;
+
+    if(!ai.abstract_state_before(p.instructions.begin()).is_bottom())
+      called.insert(f_it->first);
+  }
+
+  return called;
+}
+
+bool static_unreachable_functions(
+  const goto_modelt &goto_model,
+  const ai_baset &ai,
+  const optionst &options,
+  message_handlert &message_handler,
+  std::ostream &out)
+{
+  std::set<irep_idt> called = compute_called_functions_from_ai(goto_model, ai);
+
+  list_functions(goto_model, called, options, out, true);
+
+  return false;
+}
+
+bool static_reachable_functions(
+  const goto_modelt &goto_model,
+  const ai_baset &ai,
+  const optionst &options,
+  message_handlert &message_handler,
+  std::ostream &out)
+{
+  std::set<irep_idt> called = compute_called_functions_from_ai(goto_model, ai);
+
+  list_functions(goto_model, called, options, out, false);
+
+  return false;
```

Comment:
What's the point of this constant return value?

Reply:
Consistency with the other tasks that use that return value to set the executable's return code.

Reply:
Ok.